src/puppetlabs/puppetdb/mq_listener.clj

```diff
@@ -133,174 +156,99 @@
-  "Dump the error encountered to the log, and re-publish the message,
-  with an incremented retry counter, after a ten minute delay."
-  [{:keys [command version annotations payload] :as msg} e publish-fn]
-  (mark! (cmd-metric command version :retried))
-  (let [attempt (count (:attempts annotations))
-        id (:id annotations)
-        certname (:certname payload)
-        delay (mq/delay-property 10 :minutes)]
-    (log/error e (i18n/trs "[{0}] [{1}] Retrying after attempt {2} for {3}, due to: {4}"
-                           id command attempt certname e))
-    (time! (global-metric :retry-persistence-time)
-           (-> msg
-               (annotate-with-attempt e)
-               json/generate-string
-               (publish-fn delay)))))
-
-(defn create-message-handler
-  [publish discarded-dir message-fn]
-  (let [discard        #(dlo/store-failed-message %1 %2 discarded-dir)
-        on-discard     #(handle-command-discard % discard)
-        on-parse-error #(handle-parse-error %1 %2 discard)
-        on-fatal       #(handle-command-failure %1 %2 discard)
-        on-retry       #(handle-command-retry %1 %2 publish)]
-    (-> message-fn
-        (wrap-with-discard on-discard maximum-allowable-retries)
-        (wrap-with-exception-handling on-retry on-fatal)
-        (wrap-with-command-parser on-parse-error)
-        (wrap-with-meter (global-metric :seen))
-        (wrap-with-thread-name "command-proc"))))
+(defn mark-both-metrics!
+  "Calls `mark!` on the global and command specific metric for `k`"
+  [command version k]
+  (mark! (global-metric k))
+  (mark! (cmd-metric command version k)))
+
+(defn update-both-metrics!
+  "Calls `update!` on the global and command specific metric for `k`"
+  [command version k v]
+  (update! (global-metric k) v)
+  (update! (cmd-metric command version k) v))
+
+(defn call-with-command-metrics
+  "Invokes `f` including the related metrics updates"
+  [command version retries f]
+  (create-metrics-for-command! command version)
+
+  (mark-both-metrics! command version :seen)
+  (update-both-metrics! command version :retry-counts retries)
+
+  (mutils/multitime!
+   [(global-metric :processing-time)
+    (cmd-metric command version :processing-time)]
+
+   (let [command-result (f)]
+     (mark-both-metrics! command version :processed)
+     command-result)))
+
+;; Primarily a test hook
+(defn discard-message [message q dlo]
+  (dlo/discard-cmdref message q dlo))
+
+(defn message-handler
+  "Processes the message via (process-message msg), retrying messages
+  that fail via (delay-message msg), and discarding messages that have
+  fatal errors or have exceeded their maximum allowed attempts."
+  [q dlo delay-message process-message]
+  (fn [{:keys [certname command version delete? id] :as cmdref}]
+    (try+
+     ;; If the message is a delete?, there's no need to parse it
+     ;; below, it's only going to be removed
+     (if delete?
+       (do
+         (process-message cmdref)
```

Comment:
I don't recall now why we're processing this message on a delete. When it gets further down the stack, it just gets ignored. Maybe it'd be ok to avoid this call entirely?